src/supertux/game_object_manager.cpp

```diff
@@ -376,79 +379,233 @@ GameObjectManager::on_editor_save()
+    }
+  }
+}
+
 void
 GameObjectManager::undo()
 {
   if (m_undo_stack.empty()) return;
-  ObjectChanges& changes = m_undo_stack.back();
+  GameObjectChanges& changes = m_undo_stack.back();
 
-  for (auto& obj_change : changes.objects)
-    process_object_change(obj_change);
+  auto it = changes.objects.begin();
+  while (it != changes.objects.end())
+  {
+    try
+    {
+      process_object_change(*it);
+      it++;
+    }
+    catch (const std::exception& err)
+    {
+      log_warning << "Cannot process object change: " << err.what() << std::endl;
+      it = changes.objects.erase(it); // Drop invalid changes
+    }
+  }
 
-  m_redo_stack.push_back(std::move(changes));
+  if (!changes.objects.empty())
+  {
+    // Changes have been reversed for redo
+    m_redo_stack.push_back(std::move(changes));
+  }
   m_undo_stack.pop_back();
 }
 
 void
 GameObjectManager::redo()
 {
   if (m_redo_stack.empty()) return;
-  ObjectChanges& changes = m_redo_stack.back();
+  GameObjectChanges& changes = m_redo_stack.back();
 
-  for (auto& obj_change : changes.objects)
-    process_object_change(obj_change);
+  auto it = changes.objects.begin();
+  while (it != changes.objects.end())
+  {
+    try
+    {
+      process_object_change(*it);
+      it++;
+    }
+    catch (const std::exception& err)
+    {
+      log_warning << "Cannot process object change: " << err.what() << std::endl;
+      it = changes.objects.erase(it); // Drop invalid changes
+    }
+  }
 
-  m_undo_stack.push_back(std::move(changes));
+  if (!changes.objects.empty())
+  {
+    // Changes have been reversed for undo
+    m_undo_stack.push_back(std::move(changes));
+  }
   m_redo_stack.pop_back();
 }
 
 void
-GameObjectManager::create_object_from_change(const ObjectChange& change)
+GameObjectManager::create_object_from_change(const GameObjectChange& change, bool track_undo)
 {
   auto object = GameObjectFactory::instance().create(change.name, change.data);
-  object->m_track_undo = false;
+  object->m_track_undo = track_undo;
   object->set_uid(change.uid);
   object->after_editor_set();
   add_object(std::move(object));
 }
 
 void
-GameObjectManager::process_object_change(ObjectChange& change)
+GameObjectManager::parse_object_settings(ObjectSettings& settings, const std::string& data)
+{
+  std::istringstream stream(data);
+  auto doc = ReaderDocument::from_stream(stream);
+  auto root = doc.get_root();
+  if (root.get_name() != "supertux-game-object")
+    throw std::runtime_error("Data is not 'supertux-game-object'.");
+
+  settings.parse_state(root.get_mapping());
+}
+
+std::string
+GameObjectManager::save_object_settings_state(const ObjectSettings& settings, bool new_state)
+{
+  std::ostringstream stream;
+  Writer writer(stream);
+
+  writer.start_list("supertux-game-object");
+  if (new_state)
+    settings.save_new_state(writer);
+  else
+    settings.save_old_state(stream);
+  writer.end_list("supertux-game-object");
+
+  return stream.str();
```

Comment:
I'm quite confused about this. Assuming you call this with new_state = true, nothing gets done with the `stream` parameter. Yet you're returning `stream.str()` from the function. Does that return something usable even though the variable is initialized?

I feel like this is bad code style, but correct me if I'm wrong.

Reply:
`writer` takes a reference to `stream`. When I pass `writer` to `settings.save_new_state` effectively whatever is written to that writer is written to the stream, as it's the writer's base. So the data always goes there.

Reply:
Can you maybe change both method signatures to take a `Writer` instance? Or is that overkill?

Reply:
The old state is meant to be pre-saved and not written at the moment of calling, so it wouldn't work there.

Reply:
How about giving both a reference to `stream`? IDK...

Reply:
I just realized that doing that would complicate this more, as the writer already must be created for the "supertux-game-object" header to be added.

Reply:
Okay, even disregard my comments